### ISSUE TITLE
Figure out how to effectively use lifetimes

### ISSUE THREAD

**Issue by eddywestbrook** [GaloisInc/heapster-saw#62]
So far we have been writing our types to use the `always` lifetime everywhere, but this does not allow us to demote writes to reads in a temporary way, since all our types would need to take in lifetimes for our read permissions. We also want to be able to represent Rust lifetimes.

What is needed is a canonical way to write types that use lifetimes, along with supporting proof rules, including rules for demoting writes to reads using the relevant lifetime.

**Comment by eddywestbrook** [comment]
IDEA: there should always be an owned lifetime in scope, which is the innermost / most recently created lifetime. Any time a demotion needs to occur, it is only within that innermost lifetime.

Having a Rust lifetime in scope really just means that lifetime is guaranteed to be current during the whole of the innermost lifetime; i.e., it contains the innermost lifetime but cannot be ended during it. Intuitively, this is because a Rust lifetime can only be ended by the scope that created it. Also, only the scope that created it gets to retrieve any permissions that were stored in that lifetime, so only that scope needs an `lowned` permission for it.

**Comment by eddywestbrook** [comment]
One problem: what if a function needs to return a value whose lifetime is bound to some enclosing scope? E.g., the [Rust docs](https://doc.rust-lang.org/nomicon/lifetimes.html) show examples using an accessor function like `x = &data[0]` to borrow part of `data` into `x` for the lifetime of `x`, which is longer than the implicit call to the indexing function in `data[0]` but shorter than the lifetime of `data`. How do we represent those sorts of constraints? Solving them has to be part of how the implication solver instantiates existential lifetime variables...

**Comment by eddywestbrook** [comment]
Idea 1: the only time we need to change the lifetime of an object is when we are making a new lifetime for a borrow. So we don't need to worry about trying to change the lifetime of a pointer or other permission inside the implication prover, unless we have an `lowned` permission for the destination lifetime, in which case we will borrow the object for that lifetime.

Idea 2: accessor functions like `data[0]` above that borrow some part of a data structure should be marked somehow as creating a lifetime that lasts longer than the function. In fact, maybe the "marking" is that they add more borrows into the `lowned` permission of a lifetime that is passed to the function; e.g., the indexing function behind `data[0]` might have permission like this:

```
(l1:lifetime, ps:permlist, l2:lifetime). arg0:[l1]vec, l1:[l2]lcurrent, l2:lowned ps  -o  ret:[l2]ptr(...), l2:lowned (arg0:[l1]vec :: ps)
```

For a type like this, we always make a new lifetime for the `lowned` permission, and `l1` becomes instantiated with the lifetime on `arg0` (because we aren't going to change the lifetime of objects). The remaining question, then, is how do we know when to end `l2`. In Rust, it is when its associated variables go out of scope, so maybe that becomes a thing where we simplify away those permissions? We could also try to add support in the implication prover for looking inside `lowned` permissions in order to guess at maybe ending a lifetime to get back perms we need...

**Comment by eddywestbrook** [comment]
Modification to idea 2, above: if _any_ returned permissions use the `lowned` lifetime `l2`, then `l2` has to last longer than the function call. So it seems that the only time we need a local lifetime is if we have a function that computes a value from an input and then throws that input away, like a `length` function on a linked list. One interesting case is when an input has subcomponents at different lifetimes, in which case one of these subcomponents can be returned as part of this sort of computation that throws the input away. This matches the use case in Rust where either a shared reference to a value is passed to a function or when a mutable reference is temporarily downgraded in a function call to a shared reference.

**Comment by eddywestbrook** [comment]
The difficulty is that, other than a function type that specifically requires read permissions, we can't know when a shared reference borrow needs to happen, because in LLVM, borrowing a shared reference is the identity, i.e., the shared reference can be the same LLVM value as the pointer that it is borrowed from. We also don't know when the scope of a borrow ends.

To determine the end of a lifetime, the implication prover could be triggered to end a lifetime by trying to prove a write modality for a pointer `x` that we only have read on. It could then look through the lifetime stack for the lifetime `l` currently associated with `x` and see if ending `l` would give us back write permission to `x`.

For the beginning of a lifetime, the problem is then that we don't know the nesting of the lifetimes. If the implication prover creates one any time it needs to demote a write to a read, then every variable ends up with its own lifetime, and we end up with incomparable lifetimes... but maybe that's ok?

**Comment by eddywestbrook** [comment]
Any time the implication prover needs to demote the permissions on an object (i.e., turn `W` to `R`, either in a `ptr` or `memblock` permission or in an `array` containing one of these), it creates a lifetime, puts the current permissions in that lifetime, and then demotes, so we can promote back later. Any time the implication prover needs to promote the permissions on an object, it looks for an `lowned` lifetime containing those promoted permissions and ends that lifetime. Eventually, we will want to change `proveVarAtomicImplUnfoldOrFail` to also try to end lifetimes when it runs out of named permissions on the left it can try to unfold, but maybe that is not something we need to do right away.

Note that the above strategy for demotion assumes that the lifetime argument is an unassigned evar or a lifetime that we have `lowned` permissions on; if not, then we have to just do the demotion without saving the original permissions, like we do now.

Any time we call a function that takes in an `lowned` lifetime, we make a new lifetime, because that should be the most general solution; i.e., if we can type-check a call to a function with permissions `(l:lifetime, ps:permlist, ...). l:lowned ps, ... -o ...` using any lifetime for `l`, we could do it with a newly-created lifetime for `l`. (I haven't really checked this in detail, but I think it is true...) Note that, along with the demotion use case described above, this is the only other way we create lifetimes. In general, a function should only ever have at most one `lowned` lifetime in its input, which intuitively represents the lifetime of that function. All other lifetimes it mentions should be marked as `lcurrent` during that lifetime.

DIFFICULTY: how to handle calls to functions that require the same lifetime on multiple arguments in a context where the corresponding objects do not currently have the same lifetime. That is, if a function has input permissions `l:lowned ps, l2:[l]lcurrent, arg0:[l2]ptr(...), arg1:[l2]ptr(...), ...` and we are passing arguments for `arg0` and `arg1` that are currently in lifetimes `l0` and `l1`, respectively, we won't know how to instantiate `l2`. If one of `l0` and `l1` is nested inside the other, we could potentially set `l2` to the inner one, and coerce the other permission from the outer to the inner one, but we don't have a way to _temporarily_ coerce from an outer to an inner lifetime and then go back. That is, if we have `l0:[l1]lcurrent`, meaning `l1` is nested inside `l0`, then we could coerce an `[l0]ptr(...)` to an `[l1]ptr(...)` for the duration of a function call, but we can't then get back the `[l0]ptr(...)` after the function call is over.

**Comment by eddywestbrook** [comment]
IDEA: what if every single lifetime argument is instantiated with a freshly-created lifetime? Or, at least, every lifetime that occurs more than once?

In fact, should all lifetime arguments to a function be `lowned` instead of `lcurrent`?

**Comment by eddywestbrook** [comment]
THE PLAN: As suggested above, every lifetime argument of a function is freshly created at the function call. In general, however, function types will only have one `lowned` lifetime, which represents the lifetime of the function call; all other lifetimes will be `lcurrent`. If we want to represent that lifetime `l1` outlives `l2` in the input type of a function, written `l1:l2` in Rust, we write `l1:[l2]lcurrent`, and otherwise we write `l1:[l]lcurrent`, where `l` is the sole `lowned` lifetime of the function, to represent that `l1` outlives at least this function call. There are probably some optimizations we could do, to reduce the number of lifetimes created at function calls; e.g., when a lifetime variable is used only once, but there is no need to do that now.

NOTE: the current `SplitLifetime` rule is wrong: you can't convert `[l]ptr(...)` to `[l2]ptr(...)` with just an `l2:lowned ps` permission, because you have to know that `l` stays current while `l2` is still alive; i.e., that `l:[l2]lcurrent`. This in turn requires us to either put `l:lowned ps'` into the `lowned` permission on `l2`, or it requires an `l:[l3]lcurrent` permission for some `l3` that in turn satisfies the requirement, meaning that `l3` has to have its `lowned` permission held by `l2` or another lifetime, etc. Additionally, the `lowned` permission for a lifetime can only go into one other lifetime's `lowned` permission, so every lifetime `l` always has some "biggest" lifetime `l2`, possibly itself, holding its `lowned` permission. What this means is that, if we want to split the lifetime of an `[l]ptr(...)` permission to `l2`, we need to find the current transitive owner of `l` (the owner of the owner of etc. of `l`) and nest its `lowned` permission inside that of `l2`.

QUESTION: how do we split the lifetime of permissions containing more than one lifetime, like an array with different lifetimes on its fields, or a named permission with multiple lifetime arguments?

**Comment by eddywestbrook** [comment]
IDEA: our lifetime functors `F<l>` (like arrays with occurrences of lifetime `l` or named permissions with lifetime arguments) should always be separable into a permission of the form `P * G<l>` such that `G<l>` is equivalent to `true` when `l` is not active. Intuitively, if `F<l>` is an array or named permission that uses `l`, then `G<l>` is just those pointer permissions contained in `F<l>` that are marked with `l`, and `P` is all the parts of `F<l>` that are independent of `l`. Technically speaking, this is not quite correct, because `P` could describe the permissions to reach all the components described by `G<l>`; e.g., `P` could contain all the `next` pointers in a linked list, while `G<l>` could describe the permissions on the `data` fields. Thus I think the set `G<l>` can depend on which permission in the set `P` is chosen. This should be straightforward to define, but let's move past it for now.

We then change our `lowned` permissions to contain pairs `(F,l)` of a lifetime functor `F` and an argument `l` to `F`. The semantics of `l:lowned ((F1,l1),...,(Fn,ln))` is intuitively the permissions `G1<l1> * ... * Gn<ln>` when `l` is finished, though, again, we need to account for the fact that the `Gi<li>` permissions depend on which element of `Pi` is chosen for `Fi<li>`. The main requirement is that `F1<l> * ... * Fn<l> * l:lowned ((F1,l1), ..., (Fn,ln))` implies `F1<l1> * ... * Fn<ln>` when `l` is ended. Note that `Fi<l>` is equivalent to `Pi` when `l` is finished, so this rule is very much like saying removing the `Pi`s from the `Fi<li>`s and then re-adding the `Pi`s gets us back the `Fi<li>`s. The `F1<l> * ... * Fn<l>` thus play the role currently performed by the `minLtPerms` function... except that they don't allow read/write modalities to change back when we end a lifetime.

IDEA: we can generalize lifetime functors to lifetime + rwmodality functors `F<l,rw>`. These require that `F<l,W>` implies `F<l,R>`, meaning that `F` acts as a functor with respect to the ordering on read/write modalities, and that the two are equivalent when `l` is finished, meaning that `rw` is only used for permissions whose lifetimes are controlled by `l`. Note that there could be multiple read/write modality arguments associated with a single lifetime argument of a functor, but each read/write modality must have at most one lifetime argument it is associated with in order for this idea to work. We can check this property syntactically for most permissions, but need to record it in the definitions of named permissions. For defined and recursive permissions, we should be able to infer it automatically, and for opaque permissions we could just require there to be one lifetime argument, which intuitively controls all read/write modality arguments.

**Comment by eddywestbrook** [comment]
One more thing: we only create fresh lifetimes when proving an `l:lowned ps` or an `l:[l2]lcurrent` permission for a free variable `l`, or when demoting a `W` to a `R`. For the former, we do need to prove thoe permissions at higher priority. Note specifically that this means we will not create a fresh lifetime when just proving the `(l,rw). [l]ptr((rw,0) |-> ...)` permissions needed for pointer reads or the `(l). [l]ptr((W,0) |-> ...)` permissions needed for pointer writes

**Comment by eddywestbrook** [comment]
The steps for this change:

* Change `lowned` permissions to have the new syntax, by changing the `PermListType` to a new type called `LOwnedPerm` with the appropriate constructors; remove the old lifetime operations `inLifetime`, `containsLifetime`, and `minLtPerms`, commenting out any rules that need them; probably also remove the current `beginLifetime` and `endLifetime` functions and instructions

* Change named permissions to contain information about what `rw` arguments are contained by what lifetimes; update the commands for defining new named permissions to generate this information

* Make implication rules for beginning and ending lifetimes; change the `SplitLifetime` rule to work as described above for lifetime functors; also add a lifetime subsumption rule, that puts the `lowned` permission for one lifetime inside that of another (so we need a second form for `lowned` permissions)

* Change the implication prover to: begin fresh lifetimes when proving a lifetime perm on an existential lifetime; begin and end lifetimes as needed for pointer, array, memblock, and named permissions

* Update examples as needed, and write new ones

**Comment by eddywestbrook** [comment]
BETTER: rather than having our lifetime functors explicitly take read/write modality arguments, we can compute all the read/write modality arguments that can be weakened when we want to end a lifetime. That is, rather than have the input permissions for ending lifetime `l` be of the form

```
F1<l,args1_min> * ... * Fn<l,argsn_min> * l:lowned ((F1,args1,l1), ..., (Fn,argsn,ln))
```

we should instead of the input permissions be of the form

```
ltEndPerms(F1)<l> * ... * ltEndPerms(Fn)<l> * l:lowned ((F1,l1), ..., (Fn,ln))
```

This makes our `lowned` permissions a little simpler, but also allows us to separate weakening read/write modalities from splitting of lifetimes; i.e., it means that the `lowned` permission does not have to track what specific weakening we want to do as part of splitting the lifetime of a permission.

**Comment by eddywestbrook** [comment]
QUESTION: what is the semantics of an `lowned` permission contained in another `lowned` permission, i.e., the permission `l:lowned (l2:lowned (...))`?

ANSWER: this is just going to work differently than our lifetime functors, because the only functor that makes sense here is the constant functor that returns `l2:lowned (...)` for any lifetime argument, which would be part of the `P` part of the lifetime functor, meaning that, in our framework outlined above, it would not be held by the `lowned` permission at all. So, instead, an `lowned` permission is going to contain two different sorts of elements, lifetime functors and `lowned` permissions.

**Comment by eddywestbrook** [comment]
Wait a moment: any time we need to demote a `W` to a `R`, we will almost certainly have a lifetime which has already been determined, because this will in general only happen at function calls, where the lifetime arguments need permissions if we are to use the permissions that reference them. So function calls are the only place where we will need to create lifetimes.

**Comment by eddywestbrook** [comment]
ISSUE: when we are creating fresh lifetimes to satisfy function permissions, we cannot prove the `l:[l2]lcurrent` permissions before splitting the lifetimes of all the permissions that need to be relative to `l`, because once we prove `l:[l2]lcurrent` by putting `l:lowned ps` into the `lowned` permission for `l2`, we can no longer move other permissions into the scope of `l`. So, when proving perms, we need to first create the fresh lifetimes, then do all the non-lifetime proofs, then do all the lifetime proofs. Yuck!

**Comment by eddywestbrook** [comment]
ISSUE: when we are jumping to other basic blocks in the same function, we **don't** want to create fresh lifetimes. In fact, it is even more complicated, because we need to infer what lifetime arguments are needed for each basic block...

**Comment by eddywestbrook** [comment]
IDEA: for now, we only create fresh lifetimes when we are proving the input permissions for function calls. Eventually we might want some other mechanism for creating lifetimes for smaller scopes inside functions, but that might require hints or something...

**Comment by eddywestbrook** [comment]
ISSUE: we need some way to end a lifetime and get back write permissions held in it WITHOUT needing the associated read permissions. This is necessary for typing accessors that take in a (temporary) mutable pointer to a structure and return a mutable structure to its contents.

**Comment by eddywestbrook** [comment]
Also: we want to avoid splitting the lifetime of a write pointer with an equality perm like `ptr((W,0) |-> eq(x))` because otherwise we have to prove the pointer still points to x when we end the lifetime

**Comment by eddywestbrook** [comment]
NEW IDEA: each lifetime intuitively represents a borrow of some permissions from some larger set of permissions. A lifetime ownership permission is intuitively the permission to end that lifetime and give back those smaller permissions to get back the larger permission set. So we can model this concept by making lifetime ownership permissions as a form of implication:

`l:lowned (Pin -o Pout)`

This is the permission to end `l` and, when `l` is ended, to give back `Pin` to get back `Pout`. Because it is an implication, we have the rule

`p1 * Pin' |- p2 * Pin`  and  `p2 * Pout |- p3 * Pout'`    implies   `p1 * l:lowned(Pin -o Pout) |- p3 * l:lowned(Pin' -o Pout')`

For field permissions, we can capture the existing notion of "minimal apply" by making `Pin=ptr((R,off) |-> p)` and making `Pout=ptr((W,off) |-> p)`. That is, if we add a field permission to a lifetime, we add the read version of the field permission to `Pin` and the write version to `Pout`. If we then demote that field permission to a read, we can duplicate it and add a copy to the lifetime ownership permission, so that we do not have to keep a copy of the read permission around when we give it back to end the lifetime. This solves the problem mentioned on 2/24.

**Comment by eddywestbrook** [comment]
Building the proofs that coerce the `Pin` and `Pout` of lifetime ownership permissions is going to require some of the functionality of #61

**Comment by eddywestbrook** [comment]
It is seeming like lifetimes cannot be ended in the middle of a proof, because they might depend on other permissions that are on the stack, being used in a different part of the proof. So, instead, we only end lifetimes at the top level of the implication prover. More specifically, the implication prover first tries to prove the required permissions without ending any lifetimes, and then, if that fails, it tries to end some lifetime that could help with that proof, and recurses. Note that, in the worst case, this could be exponential, or even factorial, since (at least in the naive approach) there could be N different choices for the first lifetime to end, then N-1 for the second, etc., and at each point it tries to re-prove the required permissions and recurses. There is plenty of room to be smarter about this if it becomes necessary, but we will first try out the naive approach to see how it works in practice.

**Comment by eddywestbrook** [comment]
One remaining issue: when proving one `lowned` permission implies another, we need to substitute any equality permissions into both sides before trying to determine what extra permissions are needed to perform the local implication.
